### ISSUE TITLE
[MachineOutliner][RISC-V] DBG_VALUE affects the MBB size gate and changes outlined-function ordering

### ISSUE THREAD

**Issue by yqtian-se** [llvm/llvm-project#219176]
For RISC-V at `-Oz`, compiling the same C source with `-g` and `-g0`
produces different executable machine code.

The difference is introduced by `MachineOutliner`. Debug records cause the
pass to assign different numbers and a different order to the generated
`OUTLINED_FUNCTION_n` functions.

I have not observed a semantic difference. After ignoring function names and
function order, both objects contain the same multiset of function bodies.
This report is therefore about debug information affecting machine-code
layout, not a wrong-code miscompilation.

### Versions

The issue reproduces with:

- LLVM `llc`: `b7dc8e356b89a57b86f1571d52675327db52c506`
- Clang 24 trunk: `2e3d50b4111f324fb615c3c18d02189caa1ef62d`
- Target: `riscv32-unknown-elf`
- Architecture: `rv32gc`
- Optimization level: `-Oz`

### Reproducer

The self-contained reduced test case is attached as
`reduced-self-contained.c`. 

Compile it as follows:

```bash
clang_common=(
  --target=riscv32-unknown-elf
  -march=rv32gc
  -Oz
  -ffreestanding
  -fno-discard-value-names
  -w
  -S
  -emit-llvm
  reduced-self-contained.c
)

clang "${clang_common[@]}" -g  -o debug.ll
clang "${clang_common[@]}" -g0 -o g0.ll

llc -mtriple=riscv32-unknown-elf -verify-machineinstrs \
  -filetype=obj debug.ll -o debug.o

llc -mtriple=riscv32-unknown-elf -verify-machineinstrs \
  -filetype=obj g0.ll -o g0.o

llvm-objdump -d -w debug.o > debug.objdump
llvm-objdump -d -w g0.o    > g0.objdump

diff -u debug.objdump g0.objdump
```

### Actual behavior

After removing instruction addresses and other non-executable output from the
disassembly, the executable machine-code fingerprints are:

```text
-g:  20cdc5b540fc7e5331892976367f45823246f320360a70483a304e769094fd8b
-g0: 8cc102df48232c2dbd13b638e70964a75d0e3b80902cc4e7b2ade0f23215309a
```

The generated objects have the same function-body multiset:

```text
4dd58aefc0fbd573182726467a353c9e4b6cc9a7becabf43d00a9da4f47ba983
```

The observable difference is the ordering and numbering of the generated
`OUTLINED_FUNCTION_n` functions. The corresponding call sites are updated to
use the different numbers.

### Pass isolation

I also isolated the issue at the `MachineOutliner` boundary.

Starting from the MIR generated with debug information, I created a paired
control by removing its debug records. The ordinary MIR is identical before
running `MachineOutliner`.

Running only `machine-outliner` gives the following relation:

```text
Ordinary MIR equal before MachineOutliner:        yes
Ordinary MIR equal after MachineOutliner:         no
Natural replay matches the direct -g object:      yes
Strip-debug replay matches the direct -g0 object: yes
```

The isolated replay produces these fingerprints:

```text
Natural replay:     20cdc5b540fc7e5331892976367f45823246f320360a70483a304e769094fd8b
Strip-debug replay: 8cc102df48232c2dbd13b638e70964a75d0e3b80902cc4e7b2ade0f23215309a
```

This localizes the first ordinary machine-code difference to
`MachineOutliner`.

### Suspected cause

`MachineOutliner::populateMapper` uses `MachineBasicBlock::size()` for its
minimum-size gate:

```cpp
const unsigned MinMBBSize = 2;

if (MBB.size() < MinMBBSize)
  continue;
```

The relevant code is in `llvm/lib/CodeGen/MachineOutliner.cpp`.

A reduction of the debug records leaves one critical block containing one
ordinary `DIV` instruction and one `DBG_VALUE`:

```text
DBG_VALUE ...
DIV ...
```

With the debug record present, `MBB.size()` returns 2, so the block is passed
to `InstructionMapper::convertToUnsignedVec`. Without the debug record, the
block has size 1 and is skipped.

The block cannot produce a legal outlining range because it contains only one
ordinary instruction. However, mapping it still allocates persistent
instruction-mapping state. This changes the token numbering used for later
blocks. The changed numbering affects the traversal order of equal-benefit
outlining candidates, which changes the numbering and ordering of the
generated outlined functions.

The causal chain is:

```text
one ordinary instruction + one DBG_VALUE
                    |
                    v
          MBB.size() reaches 2
                    |
                    v
  otherwise skipped block is passed to the mapper
                    |
                    v
  unsuccessful local mapping changes global token state
                    |
                    v
 outlined-function ordering and machine code change
```

### Diagnostic patch

The following change makes the gate count only non-debug instructions:

```diff
-      if (MBB.size() < MinMBBSize) {
+      if (!MBB.sizeWithoutDebugLargerThan(MinMBBSize - 1)) {
```

With this diagnostic patch, the `-g` and `-g0` variants produce the same
ordinary output for this reproducer. The patch also closes three independently
generated source-level reproductions of the same mechanism.

### Expected behavior

Debug records such as `DBG_VALUE` should not affect the minimum-size decision
or persistent mapping state used by `MachineOutliner`.

For this test case, compiling with `-g` and `-g0` should produce the same
executable machine code after excluding debug sections and other debug-only
data.

### Attachment

- [reduced-self-contained.c](https://github.com/user-attachments/files/31511714/reduced-self-contained.c)

**Comment by lenary** [comment]
Thanks for the bug report. This sounds plausible, and potentially a bug common to other parts of the LLVM backends beyond MachineOutliner. 

That fix sounds right, please can you propose it as a PR? The usual structure of this PR will be for it to contain two commits, one which adds the testcases, and then the second which contains the fix and should also contain updates to the testcases to reflect the fixed behaviour.

Please make sure you are following the [LLVM AI policy](https://llvm.org/docs/AIToolPolicy.html).

**Comment by yqtian-se** [comment]
Thanks! I've opened #219326.

While validating the original reproducer, I found that changing the minimum block-size check alone did not fully fix the issue: an invisible debug instruction could also reset the mapper's persistent separator state and affect token numbering. The PR therefore addresses both cases, with the tests and fix split into the two commits you requested.

I have also included the AI assistance disclosure in accordance with the LLVM AI policy.

**Comment by lenary** [comment]
Thanks! I've tagged some relevant MachineOutliner people, both within RISC-V and the original pass maintainers.

I am less sure of the second fix, but only because I don't think I understand that part of the outliner. This is no evaluation of the quality of the fix.
